### PR TITLE
Authentication and QoS feature

### PR DESCRIPTION
**Pull request by ghost**
- Added username password setup methods
- Added QoS configuration validations and setup methods
- Added QoS conf class which does the abstraction for QoS related configurations
- Added protocol utility class which will encapsulate result codes for various protocols

### CODE REVIEW

File: liota/transports/mqtt.py [hunk 1]
```diff
@@ -44,29 +46,112 @@ def on_subscribe(self, client, userdata, mid, granted_qos):
 
     def on_message(self, client, userdata, msg):
         log.debug("On Message {0} {1} {2}".format(msg.topic, str(msg.qos), str(msg.payload)))
+        self.client.disconnect()
```
Comment (diwanc) [inline liota/transports/mqtt.py]: What does this line signify? self.client.disconnect()

File: liota/transports/mqtt.py [hunk 2]
```diff
@@ -44,29 +46,112 @@ def on_subscribe(self, client, userdata, mid, granted_qos):
...
         log.debug("mid: {0}".format(str(mid)))
 
-    def __init__(self, url, port):
+    def on_connect(self, client, userdata, flags, rc):
+        if rc != 0:
+            raise ValueError("Error: " + self.mqttresults.resultcodes.get(rc))
+        else:
+            log.debug("Connected with result code " + str(rc) + " : " + self.mqttresults.resultcodes.get(rc))
+
+    def __init__(self, gw_identity, TLSConf, QoSConf, url, port, keepalive, tls_support=True, enable_authentication = True):
```
Comment (diwanc) [inline liota/transports/mqtt.py]: 1) Rename QoSConf to qos_details
2) Rename TLSConf to tls_details
3) Get rid of tls_support boolean parameter. Check for tls_details presence instead
4) Ensure default values are defined for qos attributes. Check if the paho client implementation provides default values, in which case we should be good

File: liota/transports/mqtt.py [hunk 3]
```diff
@@ -44,29 +46,112 @@ def on_subscribe(self, client, userdata, mid, granted_qos):
...
+        if self.enable_authentication:
+            if not self.gw_identity.username:
+                raise ValueError("Username not found")
+            elif not self.gw_identity.password:
+                raise ValueError("Password not found")
+            else:
+                self.client.username_pw_set(self.gw_identity.username, self.gw_identity.password)
+
+        # Set QoS parameters
+        if self.QoSConf.qoslevel:
```
Comment (diwanc) [inline liota/transports/mqtt.py]: What if the qoslevel is not present?

File: liota/transports/mqtt.py [hunk 4]
```diff
@@ -44,29 +46,112 @@ def on_subscribe(self, client, userdata, mid, granted_qos):
...
+            if self.QoSConf.qoslevel>3:
+                raise ValueError("Error: Invalid QoS level")
+            elif self.QoSConf.qoslevel>0:
+                self.client.max_inflight_messages_set(self.QoSConf.inflight)
+                self.client.max_queued_messages_set(self.QoSConf.queue_size)
+                self.client.message_retry_set(self.QoSConf.retry)
+
 
+        # Connect with MQTT Broker
+        self.client.on_connect = self.on_connect
```
Comment (diwanc) [inline liota/transports/mqtt.py]: Move on_connect binding to the init method where other callback methods are registered

File: liota/utilities/protocol_utility.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
+# -*- coding: utf-8 -*-
```
Comment (diwanc) [inline liota/utilities/protocol_utility.py]: This utility is not required.
Instead leverage paho provided error code and conn ack code lookup methods. We should not attempt to re-map code and description once again. Refer https://github.com/eclipse/paho.mqtt.python/blob/master/src/paho/mqtt/client.py

File: liota/identity/qos_conf.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
+# -*- coding: utf-8 -*-
```
Comment (diwanc) [inline liota/identity/qos_conf.py]: 1) Rename QoSConf to QoSDetails
2) This file cannot be in the identity layer. Move the class to mqtt.py

File: liota/identity/tls_conf.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+#  THE POSSIBILITY OF SUCH DAMAGE.                                            #
+# ----------------------------------------------------------------------------#
+
+
+import logging
+
+log = logging.getLogger(__name__)
+
+# Encapsulates TLS configurations
+class TLSConf:
```
Comment (diwanc) [inline liota/identity/tls_conf.py]: Rename to TLSDetails

File: liota/transports/mqtt.py (hunk 1 above)
Comment (ghost) [reply liota/transports/mqtt.py]: This was for local testing purpose, will remove it

**Pull request closed without merging**